File: internal/entity/order.go
package entity

import (
	"errors"
	"time"
)

const (
	OrderSideUnknown OrderSideType = iota
	OrderSideBuy
	OrderSideSell
)

const (
	OrderKindUnknown OrderKindType = iota
	OrderKindMarket
	OrderKindLimit
)

type Order struct {
	RestingOrder
	LimitPrice   Price         `json:"limitPrice"`
	Expiry       time.Time     `json:"expiry"`
	OrderKind    OrderKindType `json:"kind"`
	OrderSide    OrderSideType `json:"side"`
	AllowPartial bool          `json:"allowPartial"`
}

type OrderNode struct {
	RestingOrder RestingOrder
	Next         NodeID
	Prev         NodeID
}

func (k OrderKindType) validate() error {
	if k != OrderKindMarket && k != OrderKindLimit {
		return errors.New("invalid order kind")
	}
	return nil
}

func (s OrderSideType) validate() error {
	if s != OrderSideBuy && s != OrderSideSell {
		return errors.New("invalid order side")
	}
	return nil
}

func (o *Order) Validate() error {
	if err := o.OrderKind.validate(); err != nil {
		return err
	}
	if err := o.OrderSide.validate(); err != nil {
		return err
	}

	if o.OriginalQty <= 0 {
		return errors.New("quantity must be positive")
	}

	if o.LimitPrice <= 0 {
		return errors.New("price must be positive")
	}
	return nil
}

// Crosses reports whether the order's limit price is aggressive enough to
// match against the opposite side's best price.
func (o *Order) Crosses(price Price) bool {
	if o.OrderSide == OrderSideBuy {
		return o.LimitPrice >= price
	}
	return o.LimitPrice <= price
}

File: src/domain/order.rs
use crate::domain::status::OrderStatus;
use serde::{Deserialize, Serialize};
use thiserror::Error;

use super::{OrderId, Price, Quantity, Sequence, UserId};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum OrderSide {
    Buy,
    Sell,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum OrderKind {
    Market,
    Limit,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RestingOrder {
    pub id: OrderId,
    pub user_id: UserId,
    pub original_qty: Quantity,
    pub open_qty: Quantity,
    pub accepted_sequence: Sequence,
    pub status: OrderStatus,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Order {
    #[serde(flatten)]
    pub resting: RestingOrder,
    pub limit_price: Option<Price>,
    pub kind: OrderKind,
    pub side: OrderSide,
}

#[derive(Debug, Clone, Copy, Error, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum OrderError {
    #[error("quantity must be positive")]
    ZeroQuantity,
    #[error("open quantity must be positive and no greater than original quantity")]
    InvalidOpenQuantity,
    #[error("limit orders require a positive price")]
    InvalidLimitPrice,
    #[error("new orders must have new status")]
    InvalidInitialStatus,
}

impl Order {
    /// Validates the order's quantity and price constraints.
    ///
    /// # Errors
    ///
    /// Returns an error when the order cannot be accepted by the matching core.
    pub fn validate(&self) -> Result<(), OrderError> {
        if self.resting.status != OrderStatus::New {
            return Err(OrderError::InvalidInitialStatus);
        }
        if self.resting.original_qty == Quantity::from(0) {
            return Err(OrderError::ZeroQuantity);
        }
        if self.resting.open_qty == Quantity::from(0)
            || self.resting.open_qty > self.resting.original_qty
        {
            return Err(OrderError::InvalidOpenQuantity);
        }
        if self.kind == OrderKind::Limit
            && self.limit_price.is_none_or(|price| price <= Price::from(0))
        {
            return Err(OrderError::InvalidLimitPrice);
        }
        Ok(())
    }

    pub fn crosses(&self, maker_price: Price) -> bool {
        match self.side {
            OrderSide::Buy => self.limit_price.is_some_and(|price| price >= maker_price),
            OrderSide::Sell => self.limit_price.is_some_and(|price| price <= maker_price),
        }
    }
}
